libsolidity/formal/SMTEncoder.cpp

```diff
@@ -3141,47 +3121,45 @@ std::set<SourceUnit const*, ASTNode::CompareByID> SMTEncoder::sourceDependencies
 	return sources;
 }
 
-void SMTEncoder::createReturnedExpressions(FunctionCall const& _funCall, ContractDefinition const* _contextContract)
+void SMTEncoder::createReturnedExpressions(FunctionDefinition const* _funDef, Expression const& _callStackExpr)
 {
-	auto funDef = functionCallToDefinition(_funCall, currentScopeContract(), _contextContract);
-	if (!funDef)
+	if (!_funDef)
 		return;
 
-	auto const& returnParams = funDef->returnParameters();
+	auto const& returnParams = _funDef->returnParameters();
 	for (auto param: returnParams)
 		createVariable(*param);
 	auto returnValues = applyMap(returnParams, [this](auto const& param) -> std::optional<smtutil::Expression> {
 		solAssert(param && m_context.knownVariable(*param), "");
 		return currentValue(*param);
 	});
-	defineExpr(_funCall, returnValues);
+	defineExpr(_callStackExpr, returnValues);
 }
 
-std::vector<smtutil::Expression> SMTEncoder::symbolicArguments(FunctionCall const& _funCall, ContractDefinition const* _contextContract)
+std::vector<smtutil::Expression> SMTEncoder::symbolicArguments(
+	FunctionDefinition const* _funDef,
+	Expression const* _calledExpr,
+	FunctionType const* _funType,
+	std::vector<ASTPointer<Expression const>> const& _arguments)
 {
-	auto funDef = functionCallToDefinition(_funCall, currentScopeContract(), _contextContract);
-	solAssert(funDef, "");
+	solAssert(_funDef, "");
+	solAssert(_funType, "");
 
 	std::vector<smtutil::Expression> args;
-	Expression const* calledExpr = &_funCall.expression();
-	auto funType = dynamic_cast<FunctionType const*>(calledExpr->annotation().type);
-	solAssert(funType, "");
-
-	std::vector<ASTPointer<Expression const>> arguments = _funCall.sortedArguments();
-	auto functionParams = funDef->parameters();
+	auto functionParams = _funDef->parameters();
 	unsigned firstParam = 0;
-	if (funType->hasBoundFirstArgument())
+	if (_funType->hasBoundFirstArgument())
 	{
-		calledExpr = innermostTuple(*calledExpr);
-		auto const& attachedFunction = dynamic_cast<MemberAccess const*>(calledExpr);
+		_calledExpr = innermostTuple(*_calledExpr);
+		auto const& attachedFunction = dynamic_cast<MemberAccess const*>(_calledExpr);
 		solAssert(attachedFunction, "");
 		args.push_back(expr(attachedFunction->expression(), functionParams.front()->type()));
 		firstParam = 1;
 	}
```

Comment:
The more I am looking at this, the more I would like to refactor this method somehow to split the treatment of the extra argument in the special case of `MemberAccess` and the conversion of actual arguments to SMT expressions.

One way that comes to my mind is to check for this optional extra argument beforehand and pass here just `std::optional<smtutil::Expression>`. Then instead of `_funDef, _calledExpr, _funType`, this method would only need the formal parameters (what we now get from `_funDef->parameters()`).

Reply:
Done